### ISSUE TITLE
Support reverse proxy injection of id_token and impersonation header

### ISSUE THREAD

**Issue by mlbiam** [kiali/kiali#3207]
**Is your feature request related to a problem? Please describe.**

The current OIDC implementation will pass the user's `id_token` to the API server, but requires that the API server is configured to use OpenID Connect.  Managed clusters, such as EKS, AKS, etc, do not support OpenID Connect.  The k8s dashboard gets around this issue by supporting reverse proxy authentication and the impersonation headers supported by kubernetes letting a reverse proxy take ownership of authentication.  This allows SSO for Kiali regardless of the type of cluster it's running on.

**Describe the solution you'd like**
I want to donate the code that would let Kiali support the same rules as the kubernetes dashboard:

1. TLS must be enabled on Kiali to support the `Authorization` header or the `Impersonate-*` headers
2.  If TLS is enabled, when the `Authorization` header is present it is passed through to the API server.  If, additionally, the `Impersonate-*` headers are present then those headers are passed through too

There's no need to validate the headers, it's the responsibility of the reverse proxy to create the correct headers and the API server to validate them.  Kiali is simply passing them through.  This makes Kiali more secure as it doesn't need a privileged service account.

**Describe alternatives you've considered**

OIDC is one alternative for OIDC enabled clusters.  This doesn't handle managed clusters though.  To handle managed clusters Kiali would need to implement token management for every cloud (ie EKS, AKS, GKE, DigitalOcean, etc...).  Kiali shouldn't gatekeep to certain platforms when a platform agnostic solution is available.

**Additional context**

Diagram of the deployment architecture:

![k8s_dashboard_impersonation](https://user-images.githubusercontent.com/8249283/93100509-3a6d3c80-f677-11ea-90c2-518dfb946d37.png)

similar issue for the k8s dashboard - https://github.com/kubernetes/dashboard/issues/3664 includes test cases

**Comment by israel-hdez** [comment]
@mlbiam We will be happy to incorporate your code into Kiali.

**Comment by mlbiam** [comment]
fantastic!  i'll get started

**Comment by GoelDeepak** [comment]
+1 I was about to file a similar enhancement and found this thread. @mlbiam looking forward for your implementation

**Comment by mlbiam** [comment]
Life's gotten in the way but hoping to get started later this week. Open to help testing?

**Comment by GoelDeepak** [comment]
> Life's gotten in the way but hoping to get started later this week. Open to help testing?

Definitely. Let me know once you have something ready.

**Comment by GoelDeepak** [comment]
@mlbiam Did you get a chance to spend sometime on this one?

**Comment by jshaughn** [comment]
@mlbiam Any updates?

**Comment by mlbiam** [comment]
finally making some progress!  I have the passthrough of the header token working:

1. Update your image to `docker.io/mlbiam/kiali:1.26`
2. In your config, change `auth.strategy` to `header`
3. Delete the `kiali` `ClusterRoleBinding` (to ensure that kiali is running without any existing privileges
4. Deploy a reverse proxy that will inject an OIDC token into the `Authorization` header (like openunison or oauth2 proxy)

Starting work on getting the impersonation headers to make it from the authentication layer down to client creation.  hopefully get that working early next week.

**Comment by mlbiam** [comment]
@GoelDeepak or anyone else is able to try it out, the above image now supports impersonation too

**Comment by GoelDeepak** [comment]
@mlbiam Thanks for working on this. I will give it a try and get back to you

**Comment by mlbiam** [comment]
@GoelDeepak not yet.  Also, don't delete the `kiali` `ClusterRoleBinding`, it causes all kind of havoc.  I'll put together a `ClusterRole` for kiali that minimzes permisions

**Comment by mlbiam** [comment]
@GoelDeepak operator is now working.  Update your `kiali-operator` deployment in the `kiali-operator` namespace to point to `docker.io/mlbiam/kiali-operator`.  Once it restarts, update your `kiali` object to look like:

```
spec:
  auth:
    strategy: header
  deployment:
    accessible_namespaces:
    - '**'
    image_name: docker.io/mlbiam/kiali
    image_pull_policy: Always
    image_pull_secrets: []
    image_version: 1.28
```

**Comment by GoelDeepak** [comment]
@mlbiam did you push `kiali-operator` image to dockerhub and is it public? I somehow can not see it. Also, are you in any open slack community channel?

**Comment by mlbiam** [comment]
@GoelDeepak yes - https://hub.docker.com/repository/docker/mlbiam/kiali-operator - i'm in k8s slack, same handle

**Comment by schroedt** [comment]
Hi @mlbiam,

I would like to test your changes, but struggel applying the same. I am utilizing Nginx as reverse Proxy. I am not that experienced with reverse proxing (e.g. hard coded port, as I do not manage getting the initial callers port extracted):
```
# DNS Update
resolver kube-dns.kube-system.svc.cluster.local;
 
# Shared memory zone
limit_req_zone $binary_remote_addr zone=limit:10m rate=2000r/m; # requests / min
limit_conn_zone $binary_remote_addr zone=addr:10m;              # Connection limit

[...]
 
server {
    location / {
    
	# Security Limits
	limit_req zone=limit burst=500 nodelay; # or delay=15;
	limit_conn addr 100;
	
        proxy_set_header     X-Real-IP $remote_addr;
        proxy_set_header     HOST $http_host;
        proxy_set_header     X-NginX-Proxy true;
 
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade; 
        proxy_set_header Connection $connection_upgrade;
        
        # Keycloak
	#proxy_set_header X-Forwarded-Host  $http_host;
 
        proxy_set_header Referer $http_referer;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
	proxy_set_header X-Forwarded-Proto $scheme;
	proxy_set_header X-Forwarded-Port 443;
 
	set $upstream redirect.example.de;
        proxy_pass https://$upstream:30000;
        proxy_redirect off;
    }
}
```

Can you provide me the exact update for `kiali-operator deployment`:
```
kubectl edit -n kiali-operator deployment.apps/kiali-operator
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
    meta.helm.sh/release-name: kiali-operator
    meta.helm.sh/release-namespace: kiali-operator
  creationTimestamp: "2021-01-11T21:12:13Z"
  generation: 1
  labels:
    app: kiali-operator
    app.kubernetes.io/instance: kiali-operator
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: kiali-operator
    app.kubernetes.io/part-of: kiali-operator
    app.kubernetes.io/version: v1.28.1
    helm.sh/chart: kiali-operator-1.28.1
    version: v1.28.1
  name: kiali-operator
  namespace: kiali-operator
  resourceVersion: "21245150"
  selfLink: /apis/apps/v1/namespaces/kiali-operator/deployments/kiali-operator
  uid: 18c4ffa1-8af5-4897-9043-c0ed840ab5b5
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.kubernetes.io/instance: kiali-operator
      app.kubernetes.io/name: kiali-operator
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      annotations:
        prometheus.io/scrape: "true"
      creationTimestamp: null
      labels:
        app: kiali-operator
        app.kubernetes.io/instance: kiali-operator
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: kiali-operator
        app.kubernetes.io/part-of: kiali-operator
        app.kubernetes.io/version: v1.28.1
        helm.sh/chart: kiali-operator-1.28.1
        name: kiali-operator
        version: v1.28.1
      name: kiali-operator
      namespace: kiali-operator
    spec:
      affinity: {}
      containers:
      - args:
        - --zap-level=info
        env:
        - name: WATCH_NAMESPACE
        - name: POD_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.name
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
        - name: OPERATOR_NAME
          value: kiali-operator
        - name: ANSIBLE_DEBUG_LOGS
          value: "true"
        - name: ANSIBLE_VERBOSITY_KIALI_KIALI_IO
          value: "1"
        - name: ANSIBLE_CONFIG
          value: /etc/ansible/ansible.cfg
        image: quay.io/kiali/kiali-operator:v1.28.1
        imagePullPolicy: Always
        name: operator
        ports:
        - containerPort: 8383
          name: http-metrics
          protocol: TCP
        - containerPort: 8686
          name: cr-metrics
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /tmp/ansible-operator/runner
          name: runner
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: kiali-operator
      serviceAccountName: kiali-operator
      terminationGracePeriodSeconds: 30
      volumes:
      - emptyDir: {}
        name: runner
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2021-01-11T21:12:30Z"
    lastUpdateTime: "2021-01-11T21:12:30Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2021-01-11T21:12:13Z"
    lastUpdateTime: "2021-01-11T21:12:30Z"
    message: ReplicaSet "kiali-operator-5f59c57ffd" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 1
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1
```

I will remove `kiali-cr`from `isito-system`before updating the deployment.

**Comment by mlbiam** [comment]
my only update the kiali operator was to update the image of the kiali operator deployment to point to the image generated from `make container-build`.

**Comment by schroedt** [comment]
Hi @mlbiam,

I tried updating deployment:

````
kubectl edit deployment.apps/kiali-operator -n kiali-operator
error: deployments.apps "kiali-operator" is invalid
deployment.apps/kiali-operator edited
````

Updates seem requiring roll-out in sequence (first `image`, than `.sec.auth.strategy: header`), as it fails applying both together.

![image](https://user-images.githubusercontent.com/6430367/104285997-ba37d600-54b4-11eb-99b1-fed0f5f52905.png)

##Update fails:
````
kubectl describe -n kiali-operator pod/kiali-operator-6484789698-jqtx2
Name:         kiali-operator-6484789698-jqtx2
Namespace:    kiali-operator
Priority:     0
Node:         kubernetes-slave02/192.168.191.66
Start Time:   Tue, 12 Jan 2021 09:01:58 +0100
Labels:       app=kiali-operator
              app.kubernetes.io/instance=kiali-operator
              app.kubernetes.io/managed-by=Helm
              app.kubernetes.io/name=kiali-operator
              app.kubernetes.io/part-of=kiali-operator
              app.kubernetes.io/version=v1.28.1
              helm.sh/chart=kiali-operator-1.28.1
              istio.io/rev=default
              name=kiali-operator
              pod-template-hash=6484789698
              security.istio.io/tlsMode=istio
              service.istio.io/canonical-name=kiali-operator
              service.istio.io/canonical-revision=v1.28.1
              version=v1.28.1
Annotations:  cni.projectcalico.org/podIP: 192.168.0.2/32
              cni.projectcalico.org/podIPs: 192.168.0.2/32
              prometheus.io/path: /stats/prometheus
              prometheus.io/port: 15020
              prometheus.io/scrape: true
              sidecar.istio.io/status:
                {"version":"e2cb9d4837cda9584fd272bfa1f348525bcaacfadb7e9b9efbd21a3bb44ad7a1","initContainers":["istio-init"],"containers":["istio-proxy"]...
Status:       Running
IP:           192.168.0.2
IPs:
  IP:           192.168.0.2
Controlled By:  ReplicaSet/kiali-operator-6484789698
Init Containers:
  istio-init:
    Container ID:  docker://70bee60838fcbd0c5195080139dfc467d3d09877800cfeb428744a22e8a92dbe
    Image:         docker.io/istio/proxyv2:1.8.1
    Image ID:      docker-pullable://istio/proxyv2@sha256:0a407ecee363d8d31957162b82738ae3dd09690668a0168d660044ac8fc728f0
    Port:          <none>
    Host Port:     <none>
    Args:
      istio-iptables
      -p
      15001
      -z
      15006
      -u
      1337
      -m
      REDIRECT
      -i
      *
      -x

      -b
      *
      -d
      15090,15021,15020
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Tue, 12 Jan 2021 09:02:08 +0100
      Finished:     Tue, 12 Jan 2021 09:02:08 +0100
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     2
      memory:  1Gi
    Requests:
      cpu:     100m
      memory:  128Mi
    Environment:
      DNS_AGENT:
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kiali-operator-token-xq89b (ro)
Containers:
  operator:
    Container ID:  docker://a8db05c4bc7eba297f37d8aeaf6996a66417b034b9327bb99d3df6b115f5422b
    Image:         docker.io/mlbiam/kiali:1.28
    Image ID:      docker-pullable://mlbiam/kiali@sha256:79aa29fc7d6a5b017c1a9b271bd886456caff56195662f8bf2f8cd90239b273e
    Ports:         8383/TCP, 8686/TCP
    Host Ports:    0/TCP, 0/TCP
    Args:
      --zap-level=info
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    2
      Started:      Tue, 12 Jan 2021 09:03:17 +0100
      Finished:     Tue, 12 Jan 2021 09:03:17 +0100
    Ready:          False
    Restart Count:  2
    Environment:
      WATCH_NAMESPACE:
      POD_NAME:                          kiali-operator-6484789698-jqtx2 (v1:metadata.name)
      POD_NAMESPACE:                     kiali-operator (v1:metadata.namespace)
      OPERATOR_NAME:                     kiali-operator
      ANSIBLE_DEBUG_LOGS:                true
      ANSIBLE_VERBOSITY_KIALI_KIALI_IO:  1
      ANSIBLE_CONFIG:                    /etc/ansible/ansible.cfg
    Mounts:
      /tmp/ansible-operator/runner from runner (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kiali-operator-token-xq89b (ro)
  istio-proxy:
    Container ID:  docker://878e0b4e05d146235eb2bfa13e9aded47c18fc9a183e7a0764d05b618fa80fd9
    Image:         docker.io/istio/proxyv2:1.8.1
    Image ID:      docker-pullable://istio/proxyv2@sha256:0a407ecee363d8d31957162b82738ae3dd09690668a0168d660044ac8fc728f0
    Port:          15090/TCP
    Host Port:     0/TCP
    Args:
      proxy
      sidecar
      --domain
      $(POD_NAMESPACE).svc.cluster.local
      --serviceCluster
      kiali-operator.$(POD_NAMESPACE)
      --proxyLogLevel=warning
      --proxyComponentLogLevel=misc:error
      --concurrency
      2
    State:          Running
      Started:      Tue, 12 Jan 2021 09:02:37 +0100
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     2
      memory:  1Gi
    Requests:
      cpu:      100m
      memory:   128Mi
    Readiness:  http-get http://:15021/healthz/ready delay=1s timeout=3s period=2s #success=1 #failure=30
    Environment:
      JWT_POLICY:                    first-party-jwt
      PILOT_CERT_PROVIDER:           istiod
      CA_ADDR:                       istiod.istio-system.svc:15012
      POD_NAME:                      kiali-operator-6484789698-jqtx2 (v1:metadata.name)
      POD_NAMESPACE:                 kiali-operator (v1:metadata.namespace)
      INSTANCE_IP:                    (v1:status.podIP)
      SERVICE_ACCOUNT:                (v1:spec.serviceAccountName)
      HOST_IP:                        (v1:status.hostIP)
      CANONICAL_SERVICE:              (v1:metadata.labels['service.istio.io/canonical-name'])
      CANONICAL_REVISION:             (v1:metadata.labels['service.istio.io/canonical-revision'])
      PROXY_CONFIG:                  {"proxyMetadata":{"DNS_AGENT":""}}

      ISTIO_META_POD_PORTS:          [
                                         {"name":"http-metrics","containerPort":8383,"protocol":"TCP"}
                                         ,{"name":"cr-metrics","containerPort":8686,"protocol":"TCP"}
                                     ]
      ISTIO_META_APP_CONTAINERS:     operator
      ISTIO_META_CLUSTER_ID:         Kubernetes
      ISTIO_META_INTERCEPTION_MODE:  REDIRECT
      ISTIO_METAJSON_ANNOTATIONS:    {"prometheus.io/scrape":"true"}

      ISTIO_META_WORKLOAD_NAME:      kiali-operator
      ISTIO_META_OWNER:              kubernetes://apis/apps/v1/namespaces/kiali-operator/deployments/kiali-operator
      ISTIO_META_MESH_ID:            cluster.local
      TRUST_DOMAIN:                  cluster.local
      DNS_AGENT:
      ISTIO_PROMETHEUS_ANNOTATIONS:  {"scrape":"true","path":"","port":""}
    Mounts:
      /etc/istio/pod from istio-podinfo (rw)
      /etc/istio/proxy from istio-envoy (rw)
      /var/lib/istio/data from istio-data (rw)
      /var/run/secrets/istio from istiod-ca-cert (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kiali-operator-token-xq89b (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  runner:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
  kiali-operator-token-xq89b:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  kiali-operator-token-xq89b
    Optional:    false
  istio-envoy:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  <unset>
  istio-data:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
  istio-podinfo:
    Type:  DownwardAPI (a volume populated by information about the pod)
    Items:
      metadata.labels -> labels
      metadata.annotations -> annotations
  istiod-ca-cert:
    Type:        ConfigMap (a volume populated by a ConfigMap)
    Name:        istio-ca-root-cert
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason     Age                From               Message
  ----     ------     ----               ----               -------
  Normal   Scheduled  84s                default-scheduler  Successfully assigned kiali-operator/kiali-operator-6484789698-jqtx2 to kubernetes-slave02
  Normal   Pulling    82s                kubelet            Pulling image "docker.io/istio/proxyv2:1.8.1"
  Normal   Created    75s                kubelet            Created container istio-init
  Normal   Pulled     75s                kubelet            Successfully pulled image "docker.io/istio/proxyv2:1.8.1" in 6.965748804s
  Normal   Started    74s                kubelet            Started container istio-init
  Normal   Pulled     51s                kubelet            Successfully pulled image "docker.io/mlbiam/kiali:1.28" in 22.646167424s
  Normal   Pulling    50s                kubelet            Pulling image "docker.io/istio/proxyv2:1.8.1"
  Normal   Pulled     46s                kubelet            Successfully pulled image "docker.io/istio/proxyv2:1.8.1" in 4.406089734s
  Normal   Created    45s                kubelet            Created container istio-proxy
  Normal   Started    45s                kubelet            Started container istio-proxy
  Normal   Pulled     31s                kubelet            Successfully pulled image "docker.io/mlbiam/kiali:1.28" in 13.614855122s
  Normal   Pulling    16s (x3 over 74s)  kubelet            Pulling image "docker.io/mlbiam/kiali:1.28"
  Normal   Pulled     6s                 kubelet            Successfully pulled image "docker.io/mlbiam/kiali:1.28" in 10.361851075s
  Normal   Created    5s (x3 over 50s)   kubelet            Created container operator
  Normal   Started    5s (x3 over 50s)   kubelet            Started container operator
  Warning  BackOff    4s (x4 over 30s)   kubelet            Back-off restarting failed container

fabiansc@Kubernetes-Master:~$ kubectl logs -f -n kiali-operator  pod/kiali-operator-6484789698-jqtx2 -c operator
flag provided but not defined: -zap-level
Usage of /opt/kiali/kiali:
  -config string
        Path to the YAML configuration file. If not specified, environment variables will be used for configuration.
```

**Comment by jmazzitelli** [comment]
> flag provided but not defined: -zap-level

You messed up your Deployment yaml. "zap-level" is an option to the [OPERATOR](https://github.com/kiali/helm-charts/blob/v1.28.0/kiali-operator/templates/deployment.yaml#L45) not to the KIALI SERVER. So your Deployment is pointing to the incorrect image.

**Comment by jmazzitelli** [comment]
I have reviewed the PRs for the helm-charts and operator - it all looks good to me. I have not run nor tested with these PRs yet so I did not give a final "green light" approval yet. But the code changes themselves all look good.

**Comment by mlbiam** [comment]
Here's how I tested:

1. Deploy Istio (obv)
2. Download my fork of the github repo (https://github.com/mlbiam/helm-charts-1)
3. Create the `kiali-operator` namespace `kubectl create ns kiali-operator`
4. Deploy the operator `helm install kiali-operator /path/to/helm-charts-1/kiali-operator -n kiali-operator --set cr.create=true --set cr.namespace=istio-system --set image.repo=docker.io/mlbiam/kiali-operator --set image.tag=1.28`
5. Once kiali is running, update the `kiali` object in `istio-system` with:

```
spec:
  auth:
    strategy: header
  deployment:
    accessible_namespaces:
    - '**'
    image_name: docker.io/mlbiam/kiali
    image_pull_policy: Always
    image_pull_secrets: []
```

this will use the image `docker.io/mlbiam/kiali:v1.28.1` (for some reason specifying the `image_version` caused all kinds of issues.

**Comment by mlbiam** [comment]
Thanks @jmazzitelli, what needs to happen to get the final "green light"?

**Comment by jmazzitelli** [comment]
> Thanks @jmazzitelli, what needs to happen to get the final "green light"?

I would like someone else to at least run it with a reverse proxy and see it work. My issue is I don't know the steps to install a reverse proxy within my test cluster. If you want to post the steps required to install a reverse proxy for use within minikube that would help testing/verifying the PRs, that would help.

**Comment by mlbiam** [comment]
@jmazzitelli the tricking thing is how to do it.  For instance I have a setup built on openunison with active directory.  i can detail how that was built.  is there an identity provider type that would work best for you?  (ie openid connect, saml2, etc)?

**Comment by jmazzitelli** [comment]
For our molecule tests that exercise the openid auth strategy, we start minikube using a hack script [`k8s-minikube.sh`](https://github.com/kiali/kiali/blob/master/hack/k8s-minikube.sh) that sets up Dex OIDC optionally via the[ `--dex-enabled true`](https://github.com/kiali/kiali/blob/v1.28.1/hack/k8s-minikube.sh#L308-L311) option:

```
hack/k8s-minikube.sh start --dex-enabled true -kv 1.18.0
```

When minikube is up after that is run, you can log into the cluster using Dex OIDC user `admin@example.com` (password: `password`). The benefit to that would be:
* (a) anyone could then easily test this by running the hack script command rather than going through a whole slew of steps to just get a cluster in a state where you can even use "header" strategy and....
* (b) I could then go ahead and implement a molecule test for this in the future (which will get run nightly during CI test runs - CI tests run against a minikube cluster that is started with that k8s-minikube.sh hack script).

So if it is possible, see if you can modify that script to add a new option to let you set up something in minikube.

**Comment by mlbiam** [comment]
@jmazzitelli is there a way to do this without virtualbox easily?

**Comment by jmazzitelli** [comment]
> @jmazzitelli is there a way to do this without virtualbox easily?

The hack script takes an option `--kubernetes-driver`:

```
  -kdr|--kubernetes-driver
      The hypervisor to use. Examples of valid values: virtualbox, hyperkit, kvm2, none.
      Only used for the 'start' command.
      Default: virtualbox
```

So you can start it however it works for you.

Run the hack script with `--help` for the full list of options available. It is a glorified wrapper around minikube, with additional features like being able to install Dex, istio itself, and bookinfo demo among other things

**Comment by mlbiam** [comment]
thank, i've got it running now with local minikube i think

**Comment by mlbiam** [comment]
finally got this working!  Here's how to run:

1.  Launch minikube with dex enabled - `./k8s-minikube.sh --dex-enabled true` ( i used kvm2)
2. Once up and running there will be an additional URL in the output `kiali reverse proxy : http://kiali-proxy.192.168.39.248.nip.io:30805` - this is pre-configured to look for kiali
3. Deploy istio - `./k8s-minikube.sh istio`
4. Deploy the kiali operator  per https://github.com/kiali/kiali/issues/3207#issuecomment-758679313
5. once running, fire up a browser and go to the `kiali reverse proxy` url from the minik8s output.  You should be prompted to login and then brought to kiali!

**Comment by jmazzitelli** [comment]
My steps to test that shows the login working:

1. Install minikube with Dex, OIDC, Istio: `k8s-minikube.sh start --dex-enabled true -kv 1.18.0; k8s-minikube.sh istio`
2. Pull the 4 PRs related to this issue to local repos
3. Build the UI: `yarn install; yarn run build`
4. Build server: `make -e CONSOLE_VERSION=local clean build test`
5. Push the locally built PR images to the local minikube: `make -e CONSOLE_VERSION=local -e CLUSTER_TYPE=minikube cluster-push`
6. Deploy operator and Kiali with the new header auth strategy: `make -e HELM_CHARTS_REPO_PULL=false -e AUTH_STRATEGY=header -e CLUSTER_TYPE=minikube operator-create kiali-create`
7. Open browser to the reverse proxy URL that k8s-minikube.sh set up: `gio open http://kiali-proxy.192.168.99.100.nip.io:30805`
8. Log into the proxy using `admin@example.com` with password `password`
9. See that you get into Kiali

![header-auth-strategy](https://user-images.githubusercontent.com/2029470/104479943-fab95180-5591-11eb-9456-f185701e2576.gif)

**Comment by jmazzitelli** [comment]
There is a minor problem we need to fix. It has to do with logout. Logout takes you to a non-functioning login screen. Somehow we need that login screen to "do something". Right now you click on "log in" button and the mouse click does nothing. At minimum it should pop up a window with a message like , "You must log in using your proxy server" or something.

To replicate:
1. Log into Kiali (see previous comment on how I did this)
2. Log out of Kiali
3. See the log in page you get redirected to - click the "Log in" button and notice it does nothing. No message or anything.

![header-logout](https://user-images.githubusercontent.com/2029470/104480848-08bba200-5593-11eb-8929-104067128c23.gif)

**Comment by jmazzitelli** [comment]
@mlbiam nil-pointer ref crash in the server PR code. See: https://github.com/kiali/kiali/pull/3568#discussion_r556752612

**Comment by jmazzitelli** [comment]
> There is a minor problem we need to fix. It has to do with logout. Logout takes you to a non-functioning login screen.

This has since been fixed in the UI PR.

**Comment by jmazzitelli** [comment]
From my point of view, it appears the last thing to fix before we can merge all of these 5 PRs is the panic caused by the nil-pointer reference. This PR is blocked until that is fixed: https://github.com/kiali/kiali/pull/3568#discussion_r556752612
The other 4 PRs have been approved for merge - we'll wait to merge those until we can confirm the nil-pointer reference panic is fixed and the openid-test molecule test passes.

**Comment by mlbiam** [comment]
i'm getting a different issue with openid.  I deployed kiali per the instructions i gave above.  When I try to login to kiali via openid I get the following error:

```
request failed (HTTP response status = 401 Unauthorized)
```

and this is what i see in the logs:

```
2021-01-14T15:51:46Z INF Not handling OpenId code flow authentication: No nonce code present. Login window timed out.
```

Here's my openid config:

```
  auth:
    openid:
      authentication_timeout: 1200
      client_id: kiali-app
      insecure_skip_verify_tls: true
      issuer_uri: https://192-168-39-31.nip.io:32000
      username_claim: email
    strategy: openid
```

the request is being done in a firefox private window.  Any ideas?

**Comment by jmazzitelli** [comment]
This is how I used to do it (obviously uncommenting the appropriate fields):

https://github.com/kiali/kiali-operator/blob/master/deploy/kiali/kiali_cr_dev.yaml#L11-L16

I will try myself again. I haven't run with this in a while (I only know the openid-test molecule test gets run every night on my minikube setup).

**Comment by jmazzitelli** [comment]
@mlbiam OK, I can replicate the nil-pointer problem. Make sure this is in your Kiali CR:

```
spec:
  auth:
    openid:
      client_id: kiali-app
      insecure_skip_verify_tls: true
      issuer_uri: https://192-168-99-100.nip.io:32000
      username_claim: email
    strategy: openid
  login_token:
    signing_key: abc12345
```

Note that my minikube is on 192.168.99.100 (thus my nip URL is what you see above) - adjust accordingly to your environment.

Now point browser to http://192.168.99.100/kiali (again, adjust accordingly to your environment) and when you click on the Login button, it takes you to the Dex login script - login as `admin@example.com` password `password` and you will see the Kiali screen with some error messages. Look in the kiali pod logs and you will see the panic stack trace.

**Comment by mlbiam** [comment]
issue isn't with new code.  i'm having the issue with the existing image too (quay.io/kiali/kiali:v1.28.1)

**Comment by jmazzitelli** [comment]
works for me with v1.28.0 - I will try with v1.28.1 now:
```
$ kubectl logs -n istio-system kiali-5f7464bbb-pslpt 
2021-01-14T17:49:19Z INF Kiali: Version: v1.28.0, Commit: bfd61ebd620ffa8dc2f412dc1caf8face221d810
2021-01-14T17:49:19Z INF Using authentication strategy [openid]
2021-01-14T17:49:19Z INF Kiali: Console version: 1.28.0
...
2021-01-14T17:49:58Z INF Kiali cache for [namespace: oauth2-proxy] started
2021-01-14T17:49:58Z INF Waiting for Kiali cache for [namespace: default] to sync
2021-01-14T17:49:58Z INF Kiali cache for [namespace: default] started
2021-01-14T17:49:58Z INF Waiting for Kiali cache for [namespace: dex] to sync
2021-01-14T17:49:58Z INF Kiali cache for [namespace: dex] started
```

**Comment by mlbiam** [comment]
this appears to be a regression.  I found this error message in https://github.com/kiali/kiali/issues/3410.  When I change the image to `quay.io/edgarhz/kiali:fix-3410` login works as expected

**Comment by jmazzitelli** [comment]
Works for me with v1.28.1:

```
$ kubectl logs -n istio-system kiali-75d97d464c-njq6s 
2021-01-14T17:53:42Z INF Kiali: Version: v1.28.1, Commit: a21e4491f1ea5a02a0e2703f8ba0e6d0b8ae65ee

2021-01-14T17:53:42Z INF Using authentication strategy [openid]
2021-01-14T17:53:42Z INF Kiali: Console version: 1.28.0
...
2021-01-14T17:56:51Z INF Waiting for Kiali cache for [namespace: oauth2-proxy] to sync
2021-01-14T17:56:51Z INF Kiali cache for [namespace: oauth2-proxy] started
2021-01-14T17:56:51Z INF Waiting for Kiali cache for [namespace: dex] to sync
2021-01-14T17:56:51Z INF Kiali cache for [namespace: dex] started
```

**Comment by jmazzitelli** [comment]
@mlbiam I cannot get this to fail with any v1.28.x version. I can successfully log in with them. Check your logs and make sure you are pulling in the images you think you are.

**Comment by mlbiam** [comment]
odd.  i could have sworn it wasn't working with v1.28.1 but now it is.  OK, i have a baseline

**Comment by mlbiam** [comment]
ok, found the issue.  was setting the token in the wrong spot with openid.  a day to track down, 30 seconds to fix :-)

**Comment by jmazzitelli** [comment]
@mlbiam I have a molecule test that shows things almost working with the new header auth strategy. See: https://github.com/mlbiam/kiali-operator/pull/1 - you should merge that into your [kiali operator PR](https://github.com/kiali/kiali-operator/pull/234).

There is a problem - something doesn't seem right to me. For an unauthenticated user, I should not expect any access into the Kiali API server to return a 200 OK success code. However, this is what happens. You can easily see this by starting Kiali in the minikube+dex setup with the new `header` auth strategy and run this simple `curl` command (my minikube IP is 192.168.99.100):

```
$ curl -kv https://192.168.99.100/kiali/api/namespaces
* About to connect() to 192.168.99.100 port 443 (#0)
*   Trying 192.168.99.100...
* Connected to 192.168.99.100 (192.168.99.100) port 443 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
* skipping SSL peer certificate verification
* SSL connection using TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
* Server certificate:
* 	subject: CN=Kubernetes Ingress Controller Fake Certificate,O=Acme Co
* 	start date: Jan 14 17:35:56 2021 GMT
* 	expire date: Jan 14 17:35:56 2022 GMT
* 	common name: Kubernetes Ingress Controller Fake Certificate
* 	issuer: CN=Kubernetes Ingress Controller Fake Certificate,O=Acme Co
> GET /kiali/api/namespaces HTTP/1.1
> User-Agent: curl/7.29.0
> Host: 192.168.99.100
> Accept: */*
> 
< HTTP/1.1 200 OK
< Server: openresty/1.15.8.2
< Date: Fri, 15 Jan 2021 01:36:49 GMT
< Content-Type: application/json
< Content-Length: 2
< Connection: keep-alive
< Vary: Accept-Encoding
< 
* Connection #0 to host 192.168.99.100 left intact
[]
```

Notice I curl the kiali API server's "get namespaces" endpoint with NO `Authorization:` header - thus, I am an unauthenticated user and I should expect this to return a 401, however, you can see here, it returns a `200 OK` with an empty list of namespaces in the response (`[]`).

Other than this, the molecule test passes all other tests.

I think we need to investigate this. I do not believe that an unauthenticated request should return a "200 OK."

You can see how the `openid` auth strategy behaves when I do this same thing. Deploy the same setup but this time use `openid` strategy and try the curl again - this is what you see (notice the 401 and "Unauthorized" message):

```
$ curl -kv https://192.168.99.100/kiali/api/namespaces
* About to connect() to 192.168.99.100 port 443 (#0)
*   Trying 192.168.99.100...
* Connected to 192.168.99.100 (192.168.99.100) port 443 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
* skipping SSL peer certificate verification
* SSL connection using TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
* Server certificate:
* 	subject: CN=Kubernetes Ingress Controller Fake Certificate,O=Acme Co
* 	start date: Jan 14 17:35:56 2021 GMT
* 	expire date: Jan 14 17:35:56 2022 GMT
* 	common name: Kubernetes Ingress Controller Fake Certificate
* 	issuer: CN=Kubernetes Ingress Controller Fake Certificate,O=Acme Co
> GET /kiali/api/namespaces HTTP/1.1
> User-Agent: curl/7.29.0
> Host: 192.168.99.100
> Accept: */*
> 
< HTTP/1.1 401 Unauthorized
< Server: openresty/1.15.8.2
< Date: Fri, 15 Jan 2021 01:45:50 GMT
< Content-Type: text/plain; charset=utf-8
< Content-Length: 13
< Connection: keep-alive
< Vary: Accept-Encoding
< X-Content-Type-Options: nosniff
< 
Unauthorized
```

**Comment by jmazzitelli** [comment]
@mlbiam I believe I fixed the error mentioned above (about an unauthenticated user always seeing a 200 OK rather than a 401). Please review this PR that can be merged into your server-side PR: https://github.com/mlbiam/kiali/pull/1

**Comment by mlbiam** [comment]
@jmazzitelli great catch!  verified by the issue and your fix locally and merged the prs

**Comment by jmazzitelli** [comment]
All molecule tests now pass. If possible, I would like want one final code review by @israel-hdez since he's most familiar with the server and UI code dealing with auth. For myself, I'm OK with all the PRs and am +1 to merge them.

Molecule tests:
```
              accessible-namespaces-test... success [6m 11s]
     affinity-tolerations-resources-test... success [6m 1s]
                                api-test... success [4m 23s]
                      config-values-test... success [7m 24s]
                  default-namespace-test... success [2m 54s]
                        header-auth-test... success [2m 52s]
                            metrics-test... success [5m 21s]
                     null-cr-values-test... success [3m 3s]
                only-view-only-mode-test... success [3m 45s]
                             openid-test... success [2m 47s]
                   os-console-links-test... skipped
                              roles-test... success [8m 41s]
                    rolling-restart-test... success [3m 47s]
                              token-test... success [2m 51s]
```

**Comment by jmazzitelli** [comment]
@mlbiam I was afraid this was going to happen - the server-side PR got conflicts. Can you fix those? Only in go.[sum,mod].
Once you fix those, I'm going to merge. I was going to wait for more reviews, but I'm not going to wait anymore. Just fix that PR and then once done, I'm going to press the Big Green Buttons on all of these PRs.

**Comment by jmazzitelli** [comment]
I have a diff that should fix the conflicts - see the server side PR (for some reason, when I tried to create a PR-to-the-PR it didn't look right - so I just posted a diff - should be easy to apply). See: https://github.com/kiali/kiali/pull/3568#issuecomment-762252134

**Comment by mlbiam** [comment]
i merged the PR and made a minor update.  I accepted all incoming changes.

**Comment by jmazzitelli** [comment]
Thanks @mlbiam. All merged!

**Comment by mlbiam** [comment]
awesome!  Thanks @jmazzitelli for your help in getting this merged!

**Comment by jmazzitelli** [comment]
This appears to have broken openshift auth. Currently investigating. See: https://github.com/kiali/kiali/issues/3591

**Comment by jmazzitelli** [comment]
We might have to revert if this doesn't get fixed - the PRs to revert are:

* https://github.com/kiali/helm-charts/pull/29
* https://github.com/kiali/kiali-operator/pull/234
* https://github.com/kiali/kiali-ui/pull/2048
* https://github.com/kiali/kiali/pull/3568
* https://github.com/kiali/kiali.io/pull/340
* https://github.com/kiali/kiali/pull/3590

**Comment by jmazzitelli** [comment]
I think I may have fixed it : https://github.com/kiali/kiali/pull/3594/files
Performing more tests now

**Comment by jmazzitelli** [comment]
Re-closing this issue. No need to revert - but this PR is required: https://github.com/kiali/kiali/pull/3594 (this is issue https://github.com/kiali/kiali/issues/3591)
